ament_cmake_cppcheck/cmake/ament_cmake_cppcheck_lint_hook.cmake

```diff
@@ -33,18 +33,36 @@ if(_source_files)
   if(NOT CMAKE_VERSION VERSION_LESS "3.7.0")
     get_directory_property(_build_targets DIRECTORY ${CMAKE_SOURCE_DIR} BUILDSYSTEM_TARGETS)
     foreach(_target ${_build_targets})
-      get_property(_include_dirs
+      # Check if INCLUDE_DIRECTORIES is defined first in case the target is an INTERFACE
+      get_property(_include_dirs_defined
         TARGET ${_target}
         PROPERTY INCLUDE_DIRECTORIES
+        DEFINED
       )
 
+      if(_include_dirs_defined)
+        get_property(_include_dirs
+          TARGET ${_target}
+          PROPERTY INCLUDE_DIRECTORIES
+        )
+      else()
+        # Target might be an interface
+        get_property(_include_dirs
+          TARGET ${_target}
+          PROPERTY INTERFACE_INCLUDE_DIRECTORIES
+        )
+      endif()
+
       # Only append include directories that are from the package being tested
       # This accomplishes two things:
       #     1. Reduces execution time (less include directories to search)
       #     2. cppcheck will not check for errors in external packages
       foreach(_include_dir ${_include_dirs})
-        string(REGEX MATCH "^${CMAKE_SOURCE_DIR}.*" _is_match ${_include_dir})
-        if(_is_match)
+        # Check if include directory is a subdirectory of the source directory
+        string(REGEX MATCH "^${CMAKE_SOURCE_DIR}/.*" _is_subdirectory ${_include_dir})
+        # Check if include directory is part of a generator expression (e.g. $<BUILD_INTERFACE:...>)
+        string(REGEX MATCH "^\\$<.*:${CMAKE_SOURCE_DIR}/.*>$" _is_genexp_subdirectory "${_include_dir}")
```

Comment:
The `CMAKE_SOURCE_DIR` might contain characters which have special meaning in a regex. Unfortunately afaik there is no function in CMake to escape a string for that use case (see https://gitlab.kitware.com/cmake/cmake/issues/18409).

Reply:
I think the real world impact of this issue is narrowly scoped enough that the TODO in 600fa4c covers it.

A robust string library could probably get around the need for regexp here. Comparing the initia substring of the include directory to the CMAKE_SOURCE_DIR might be sufficient in the usual case but I don't have any idea yet what to do about the Generator Expressions.